libcomposefs/lcfs-writer-erofs.c

```diff
@@ -881,7 +881,7 @@ static int write_erofs_inode_data(struct lcfs_ctx_s *ctx, struct lcfs_node_s *no
 		i.i_uid = lcfs_u32_to_file(node->inode.st_uid);
 		i.i_gid = lcfs_u32_to_file(node->inode.st_gid);
 		i.i_mtime = lcfs_u64_to_file(node->inode.st_mtim_sec);
-		i.i_mtime_nsec = lcfs_u64_to_file(node->inode.st_mtim_nsec);
+		i.i_mtime_nsec = lcfs_u32_to_file(node->inode.st_mtim_nsec);
```

Comment:
Ah yes, good old C automatic integer promotion rules, here to blow off our feet.

So I tested this, and it finds the problem:

```
diff --git a/libcomposefs/lcfs-internal.h b/libcomposefs/lcfs-internal.h
index f9cab47..749dbe6 100644
--- a/libcomposefs/lcfs-internal.h
+++ b/libcomposefs/lcfs-internal.h
@@ -70,10 +70,10 @@ static inline uint32_t lcfs_u32_to_file(uint32_t val)
 	return htole32(val);
 }
 
-static inline uint64_t lcfs_u64_to_file(uint64_t val)
-{
-	return htole64(val);
-}
+#define lcfs_u64_to_file(v) ({ \
+	_Static_assert(sizeof(v) == sizeof(uint64_t)); \
+	htole64(v); \
+})
 
 static inline uint16_t lcfs_u16_from_file(uint16_t val)
 {
```

Opinions?

Reply:
This may hit problematic cases where we e.g. pass an int to u64_to_file(), but I'm fine with having those explicitly add a cast as needed.

Reply:
https://github.com/containers/composefs/pull/192